src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -259,22 +262,27 @@ private void OnPartialValidationSucceeded(ChainedHeaderBlock chainedHeaderBlock)
                         lock (this.treeLock)
                         {
                             ConnectNewHeadersResult connectNewHeaders = this.chainedHeaderTree.ConnectNewHeaders(peer.Connection.Id, headers);
-                            blocksToDownload.Add(connectNewHeaders);
+
+                            if (connectNewHeaders.DownloadTo != null)
+                            {
+                                blocksToDownload.Add(connectNewHeaders);
+                            }
                         }
                     }
                 }
 
                 foreach (ConnectNewHeadersResult newHeaders in blocksToDownload)
                 {
-                    this.DownloadBlocks(newHeaders.ToHashList(), this.ProcessDownloadedBlock);
+                    this.DownloadBlocks(newHeaders.ToHashArray().ToList(), this.ProcessDownloadedBlock);
```

Comment:
How about changing the signature of download blocks to accept an array so that we don't need to convert to list?